tmva/sofie/inc/TMVA/ROperator_Reduce.hxx

```diff
@@ -0,0 +1,156 @@
+   fkeepdims(keepdims), fAttrAxes(attrAxes), fNX(UTILITY::Clean_name(nameX)), fNY(UTILITY::Clean_name(nameY)) {
+      fReduceOpMode = Op;
+   }
+
+   // type of output given input
+   std::vector<ETensorType> TypeInference(std::vector<ETensorType> input){
+      return input;
+   }
+
+   // shape of output tensors given input tensors
+   std::vector<std::vector<size_t>> ShapeInference(std::vector<std::vector<size_t>> input){
+      auto ret = input; //suggest copy to compiler
+      ret[0][fAttrAxes] = 1;
+      return ret;
+   }
+    void Initialize(RModel& model){
+
+        fUseSession = model.UseSession();
+
+        if (model.CheckIfTensorAlreadyExist(fNX) == false){   //input must be a graph input, or already initialized intermediate tensor
+            throw std::runtime_error("TMVA SOFIE Reduce Op Input Tensor " + fNX + " is not found in model");
+        }
+        fShapeX = model.GetTensorShape(fNX);
+         // find shape of Y and add it in the list of intermediate tensors
+         fShapeY = ShapeInference({fShapeX})[0];
+         model.AddIntermediateTensor(fNY, model.GetTensorType(fNX), fShapeY);
+
+    }
+
+    std::string Generate(std::string OpName){
+      OpName = "op_" + OpName;
+      if (fShapeX.empty() || fShapeY.empty()) {
+         throw std::runtime_error("TMVA SOFIE Reduce Op called to Generate without being initialized first");
+      }
+
+      size_t outputLength = TMVA::Experimental::SOFIE::ConvertShapeToLength(fShapeY);
+
+      auto inputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeX);
+      auto outputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeY);
+
+      // write here according to size of shape
+      // in generation code can be done automatically
+      // i0 =  i / s0 ; i1 = (i % s0) / s1 ; i2 = ( (i % s0) % s1 ) / s2 and so on
+      // and we have for the inverse
+      // i = i0 * s0 + i1 * s1 + i2 * s2 + i3 * s3 ....
+
+      // don't need to divide by last stride s[n-1] since it is 1 by definition
+
+      std::stringstream out;
+      out << "\n//----  operator " << Name() << "  " << OpName << "\n";
+      out << SP << "for (size_t i = 0; i < " << outputLength << "; i++) {\n";
+
+      size_t dim = fShapeX.size();   // this is the input dimension (e.g. 2, 3 or 4 or more)
+
+      // here we find output indices
+      out << SP << SP << "size_t idx_0 = i / " << outputStrides[0] << ";\n" ;
+      out << SP << SP << "size_t itmp = i;\n";
+      for (size_t k = 1; k < dim; k++) {
+         out << SP << SP << "itmp = itmp % " << outputStrides[k-1] << ";\n" ;
+         if (k < dim-1)
+            out << SP << SP << "size_t idx_" << k << " = itmp / " << outputStrides[k] << ";\n" ;
+         else
+           // to avoid division by 1 which is outputStrides[dim-1]
+           out << SP << SP << "size_t idx_" << k << " = itmp;\n";
+      }
+
+      // compute reduction
+
+      out << SP << SP << "float sum = 0;\n";
```

Comment:
In case of the 'ReduceProd` operator we should have the initial sum equal to 1. 
Do:
```
if(fReduceOpMode == ReduceProd)
  out << SP << SP << "float sum = 1;\n";
else 
  out << SP << SP << "float sum = 0;\n";
```

Reply:
Done with the changes sir. All the tests pass and now PR is ready to be merged.